tools/parsergen/src/parsergen/resolver.py

```
@@ -0,0 +1,228 @@
+        if len(symbol.arguments) > len(production.parameters):
+            bag.add(_diagnostic("GR003", symbol.span, symbol.name))
+        return ResolvedNonterminal(symbol.name, symbol.arguments, symbol)
+    raise TypeError(type(symbol))
+
+
+def _diagnostic(
+    code: str,
+    span: SourceSpan,
+    symbol: str,
+    *,
+    related: tuple[RelatedLocation, ...] = (),
```

Ruff (0.16.1)

[error] Undefined name `RelatedLocation`

(F821)